examples/ttn-otaa-network-time/ttn-otaa-network-time.ino

```diff
@@ -0,0 +1,293 @@
+            break;
+        /*
+        || This event is defined but not used in the code. No
+        || point in wasting codespace on it.
+        ||
+        || case EV_RFU1:
+        ||     Serial.println(F("EV_RFU1"));
+        ||     break;
+        */
+        case EV_JOIN_FAILED:
+            Serial.println(F("EV_JOIN_FAILED"));
+            break;
+        case EV_REJOIN_FAILED:
+            Serial.println(F("EV_REJOIN_FAILED"));
+            break;
+        case EV_TXCOMPLETE:
+            Serial.println(F("EV_TXCOMPLETE (includes waiting for RX windows)"));
+            if (LMIC.txrxFlags & TXRX_ACK)
+              Serial.println(F("Received ack"));
+            if (LMIC.dataLen) {
+              Serial.print(F("Received "));
+              Serial.print(LMIC.dataLen);
+              Serial.println(F(" bytes of payload"));
+            }
+            // Schedule next transmission
+            os_setTimedCallback(&sendjob, os_getTime() + sec2osticks(TX_INTERVAL), do_send);
+            break;
+        case EV_LOST_TSYNC:
+            Serial.println(F("EV_LOST_TSYNC"));
+            break;
+        case EV_RESET:
+            Serial.println(F("EV_RESET"));
+            break;
+        case EV_RXCOMPLETE:
+            // data received in ping slot
+            Serial.println(F("EV_RXCOMPLETE"));
+            break;
+        case EV_LINK_DEAD:
+            Serial.println(F("EV_LINK_DEAD"));
+            break;
+        case EV_LINK_ALIVE:
+            Serial.println(F("EV_LINK_ALIVE"));
+            break;
+        /*
+        || This event is defined but not used in the code. No
+        || point in wasting codespace on it.
+        ||
+        || case EV_SCAN_FOUND:
+        ||    Serial.println(F("EV_SCAN_FOUND"));
+        ||    break;
+        */
+        case EV_TXSTART:
+            Serial.println(F("EV_TXSTART"));
+            break;
+        default:
+            Serial.print(F("Unknown event: "));
+            Serial.println((unsigned) ev);
+            break;
+    }
+}
+
+uint32_t userUTCTime; // Seconds since the UTC epoch
+
+// Utility function for digital clock display: prints preceding colon and
+// leading 0
+void printDigits(int digits) {
+    Serial.print(':');
+    if (digits < 10) Serial.print('0');
+    Serial.print(digits);
+}
+
+void user_request_network_time_callback(void *pVoidUserUTCTime, int flagSuccess) {
+    // Explicit conversion from void* to uint32_t* to avoid compiler errors
+    uint32_t *pUserUTCTime = (uint32_t *) pVoidUserUTCTime;
+
+    // A struct that will be populated by LMIC_getNetworkTimeReference.
+    // It contains the following fields:
+    //  - tLocal: the value returned by os_GetTime() when the time
+    //            request was sent to the gateway, and
+    //  - tNetwork: the seconds between the GPS epoch and the time
+    //              the gateway received the time request
+    lmic_time_reference_t lmicTimeReference;
+
+    if (flagSuccess != 1) {
```

Comment:
I don't understand this condition; how can flagSuccess already be set at this point to a value?

Reply:
Not sure I understand the question. It's an input parameter to this function. The caller computes it and passes it in.

Reply:
After some analysis (but sure not yet enough) of the code i still don't fully understand the control flow and don't oversee in what situation this condition is called. As far as i see it's called if 
1. a time request is ongoing, and `LMIC.txDeviceTimeReqState != lmic_RequestTimeState_success` (= time request not successful)
2. no time request is ongoing, and  `LMIC.txDeviceTimeReqState != lmic_RequestTimeState_idle`
(= improper state)

So far so good. But from the .ino example code i always get the error message "USER CALLBACK: Not a success" from the user callback routine, meaning that it is always callled with flagSuccess != 1.

Something is not working in my environment. Have to dive in deeper in the controlflow by firing up my JTAG debugger.

Reply:
Turns out:

<lmic.c>
```
#if LMIC_DEBUG_LEVEL > 0
                LMIC_DEBUG_PRINTF("%"LMIC_PRId_ostime_t": MAC command DeviceTimeAns received: seconds_since_gps_epoch=%"PRIu32", fractional_seconds=%d\n", os_getTime(), LMIC.netDeviceTime, LMIC.netDeviceTimeFrac);
#endif
```
I never see this message, so i assume i never get a MAC cmd DeviceTimeAns from the TTN network.

Reply:
Actually, if the standard packet_forwarder is used, the network would answer with a `timeSinceGpsEpoch` equal to 0. So I think he'd be able to see the `"MAC command DeviceTimeAns received"` message on the Serial connection.

Another possiblity is that TTN is not supporting the `DeviceTimeReq` commands, which are present in LoRaWAN specs 1.03 and later. However I have't used TTN myslef, so I'm not sure.

But @cyberman54 is right: if logging is enabled and he the `"MAC command DeviceTimeAns received"` message doesn't show up, I suspect the gateway is not sending it.
 @cyberman54, can I ask you to provide the `packet_forwarder` version you're using, and maybe some logs from the gateway?

Reply:
No, the code was not tested in TTN directly.

Also, those timestamps you see in the JSON you posted are _not_ the timestsamps required. The timestamp required is named `"tmms"`. 

I wrote a more in-depth explanation of why it is needed [here](https://github.com/kersing/packet_forwarder/issues/25).

The only `packet_forwarder` implementation which sends the `"tmms"` field that I know of is the one reported in the "Testing" section of this PR's description.

Reply:
I checked my gateway. It runs the semtech Lora-Net packet forwarder. In the `rxpk` packets on tcp Port 1700 i can see the `"tmms"` fields, going out to the network server of TheThingsNetwork:

`JSON up: {"rxpk":[{"tmst":352191604,"time":"2018-11-14T19:51:03.700521Z","tmms":1226260282700,"chan":1,"rfch":1,"freq":868.300000,"stat":1,"modu":"LORA","datr":"SF7BW125","codr":"4/5","lsnr":10.0,"rssi":-51,"size":17,"data":"QCUkASaAvhYBckgyBez8Uvs="}]}`

Reply:
@frazar @cyberman54 thanks for these investigations.

Reply:
@frazar The #define is enabled, sure.

Reply:
Better check whether this is supported on TTN; and if so, what you need to report from your gateway to enable it. The core network has to add the MAC to the downlink; it's not a gateway function, because the gateway doesn't have the keys. If @frazar is right that it's supported, but only if the core network thinks you have GPS, you need to make sure all that is properly provisioned. It's also possible that they don't support this yet in V2.

Reply:
This how a time request and 1 byte send cycle looks in log, using TTN:

```
> [I][cyclic.cpp:92] do_timesync(): Network time request scheduled
> [I][button.cpp:10] readButton(): Button pressed
> [I][lorawan.cpp:398] lora_enqueuedata(): 1 bytes enqueued for LORA interface
> 2954006: engineUpdate, opmode=0x808
> [I][lorawan.cpp:271] onEvent(): EV_TX START
> 2954074: engineUpdate, opmode=0x888
> 2954139: TXMODE, freq=867100000, len=15, SF=9, BW=125, CR=4/5, IH=0
> [I][lorawan.cpp:334] lora_send(): 1 byte(s) sent to LoRa
> 3014874: setupRx1 txrxFlags 0x1 --> 01
> start single rx: now-rxtime: 3
> 3014972: RXMODE_SINGLE, freq=867100000, SF=9, BW=125, CR=4/5, IH=0
> rxtimeout: entry: 3041185 rxtime: 3014965 entry-rxtime: 26220 now-entry: 4 rxtime-txend: 50468
> 3064811: setupRx2 txrxFlags 0x1 --> 02
> start single rx: now-rxtime: 4
> 3064929: RXMODE_SINGLE, freq=869525000, SF=9, BW=125, CR=4/5, IH=0
> rxtimeout: entry: 3116185 rxtime: 3064921 entry-rxtime: 51264 now-entry: 5 rxtime-txend: 100424
> 3116310: processRx2DnData txrxFlags 0x2 --> 00
> 3332689: processDnData txrxFlags 00 --> 20
> [W][lorawan.cpp:432] user_request_network_time_callback(): Network time request callback error
> [I][lorawan.cpp:271] onEvent(): EV_TX COMPLETE
> 3333021: engineUpdate, opmode=0x800
```

Reply:
Turns out: I don't get any answer from the network. After uplink with payload and MAC command DeviceTimeReq nothing happenes, no downlink follows. If i force a downlink by using a confirmed uplink, i get the ack, but not MAC command DeviceTimeAns comes with the downlink. So i assume it's a network related problem, probably TTN does not support DeviceTimeReq yet. I will ask in TTN forum.

Reply:
@frazar @terrillmoore Turns out: TTN does NOT support DeviceTimeReq on it's current v2 stack, as stated [here.](https://www.thethingsnetwork.org/forum/t/mac-command-devicetimereq-supported/19785/4)
So no need to dive in deeper here. It's a network issue. Awaiting v3...

Reply:
I see! Thank you for pointing it out, I'll add this to the PR description. 
We really need to add some proper documentation though..

Reply:
I have a new version of the docs that is going in as part of #95. My fix for #95 ran into a snag... so it's delayed. I'll make a new issue just for the doc file, and push changes on that so that @frazar @cyberman54 others can review.

Reply:
I created a branch (issue168) and a pull request (#179) to update the documents. @frazar @cyberman54 others, please review.

Reply:
I reviewed it, no comments.

Reply:
@cyberman54 thanks